src/resourceIndex.js

```diff
@@ -186,26 +193,40 @@ class ListResources {
 
 
   get data() {
-    const _resources = resources?.[this.sourceId]?.[this.resourceType];
-    if (!_resources) {
+    const _resources = Object.entries(resources?.[this.sourceId]?.[this.resourceType] || [])
+      .filter((el) => this.filterResources(el)); // fat-arrow to avoid rebinding `this`
+
+    if (!_resources.length) {
       throw new NotFound(`No resources exist for the provided source-id / resource-type`);
     }
-    if (this.resourceType !== 'dataset') {
-      throw new InternalServerError(`Resource listing is currently only implemented for datasets`);
-    }
-    const pathVersions = Object.fromEntries(
-      Object.entries(_resources).map(([name, data]) => {
-        return [name, data.versions.map((v) => v.date)];
-      })
-      .filter((d) => this.sourceId==='core' ? this.coreDatasetFilter(d) : true)
-    )
-    const d = {}
-    d[this.resourceType] = {}
-    d[this.resourceType][this.sourceId] = {
-      pathVersions,
-      pathPrefix: this.pathPrefixBySource(this.sourceId)
-    }
-    return d;
+
+    const result = {
+      [this.resourceType]: {
+        [this.sourceId]: {
+          pathPrefix: this.pathPrefixBySource(this.sourceId)
+        }
+      }
+    };
+
+    // Use conditional logic to select appropriate behaviour as it's simpler to understand
+    // than splitting behaviour across multiple classes (in this case, anyway)
+    if (this.resourceType === 'dataset') {
+      result[this.resourceType][this.sourceId].pathVersions = Object.fromEntries(_resources.map(_formatDatasets));
+    } else if (this.resourceType === 'intermediate') {
+      if (this.groupHistory) {
+        result[this.resourceType][this.sourceId].pathVersions = Object.fromEntries(
+          _resources
+            .filter(([name,]) => name.split('/')[0]===this.groupHistory)
+            .map(_formatIntermediates)
+        );
+      } else {
+        result[this.resourceType][this.sourceId].latestVersions = Object.fromEntries(_resources
+          .map(_formatIntermediates)
+          .map(_onlyLatestIntermediates)
+        );
+      }
+    }
+    return result;
```

Comment:
Suggested refactor:

```suggestion
    // Use conditional logic to select appropriate behaviour as it's simpler to understand
    // than splitting behaviour across multiple classes (in this case, anyway)
    let key;
    let entries;
    if (this.resourceType === 'dataset') {
      key = "pathVersions";
      entries = _resources.map(_formatDatasets);
    } else if (this.resourceType === 'intermediate') {
      if (this.groupHistory) {
        key = "pathVersions";
        entries = _resources
          .filter(([name,]) => name.split('/')[0]===this.groupHistory)
          .map(_formatIntermediates);
      } else {
        key = "latestVersions"
        entries = _resources
          .map(_formatIntermediates)
          .map(_onlyLatestIntermediates);
      }
    }

    result[this.resourceType][this.sourceId][key] = Object.fromEntries(entries);

    return result;
```

This extracts all (most?) of the duplicated boilerplate into the single line at the end, and IMO makes it easier to understand what the conditional logic is switching across.